### ISSUE TITLE
Discuss the advantage of C++-based ploting over Python-based ploting

### ISSUE THREAD

**Issue by yipengsun** [umd-lhcb/plot_scripts#3]
After more thoughts, I'm no longer sure about use this library for most of the plotting.

The big-picture claim is that Python is _slow for computation_, and should be avoided when we need to compute purely in Python. However I don't think plotting (and making cuts) are mostly computing-intensive (see the profiling at umd-lhcb/pyTuplingUtils#4: most of the time was spent on IO), so I don't think why we can't use Python (it's much easier to trains someone to use Python, and we can use undergrads to do some jobs with Python more easily).

I think the main advantages of this projects are:

1. Written in C++, so it can include all existing C++ code (for example, maybe cuts need to be defined only once)
2. Speed. I think there's no double that C++ code is faster than Python, but I'd like to benchmark and quantify the advantage here
3. Have very nice style already defined

The disadvantages are:

1. Cannot discover the tree structure on its own, having to rely on external YAML file.
2. Harder to change the inner mechanisms, because everything is made in-house. The Python approach uses standard libraries and should be easier to change according to our needs.

I propose two routes:

## Route 1
This is the preferred route, because this seems to save time.

We'll try to integrate this project into our current `lhcb-ntuples-gen` workflow by:

- [ ] Making the `makefile` importable, so that we can use this project nested any C++ project.
- [ ] Making the tree-structure YAML file as a cli argument, and generate the tree structure at compile time (when generating plotting executable)
- [ ] Making the ntuple as a cli argument, so we don't rely on hard-coded path for storing ntuples.

## Route 2
We need to benchmark the actual performance of the `uproot` + `matplotlib` route, and see if the performance loss is acceptable.

The advantages of the Python routes are:
1. Already have a parser for arbitrary selection string, which is very easily extensible
2. More friendly to newcomers, given that no compiler is needed (Note that this is a pure Python approach, unlike the `pyBabyMaker` which needs compilation).
3. The usage of `uproot` enables us to discover the tree structure of ntuples at runtime, **NOT** compilation time.
4. Better for the undergrads, because `matplotlib` has more use cases.

The disadvantages of Python route are the advantage of C++ route.

If we decide to go this route, we need to:
- [ ] Reimplement the plotting theme

I should mention that even though we have only 1 TODO for the Python route, the objective is likely harder than all 3 TODOs in the C++ route combined.

**Comment by yipengsun** [comment]
Anyway, I think Alex should be trained on this first, and see how easy it is to use this library, and if he complains.

But I don't think we should kill the Python route, unless we have some serious performance hit, due to the advantages listed above.

**Comment by manuelfs** [comment]
Thanks for sharing your thoughts Yipeng. Actually, the main reason to use `plot_scripts` is not speed, but simply that it already has a lot of the functionality that we'll need and it has been proven to work as it was employed to publish several papers already. Also, the inner mechanisms being customs actually give a lot of flexibility.

I am actually not so sure of the speed advantage, as I had coded an initial version of this package that was C++-fast, but when the project grew and became more flexible with `NamedFunc`s, it also became much slower.

It also has other disadvantages that you've mentioned before, like problems with compilation when branches that have been used in custom functions change name and in other cases.

Despite all, it is a package that has shown its value, and we will use it for now. The python route can remain on the table, but unless there are strong reasons of a tool limiting the workflow, we will not spend time and resources developing something new. Measuring `RD(*)` is a very large endeavor and to pull it off in a reasonable amount of time we will need to focus on the physics and on the tools that are severely needed, not on tools that may be marginally better.

Since we should not be distracted by this issue for now, we will use it how it has been used in the past, not integrating it with `lhcb-ntuple-gen`, and we can revisit this issue in 3 weeks.

**Comment by yipengsun** [comment]
I agree with the idea that we don't have time to fully discover a marginally better solution, but I disagree with the idea that we should use this library as-is.

Given that this project is mostly self-contained, the change in `makefile` should be minimal. Actually, all I need to do is probably make the header and implementation discoverable and I can use the project anywhere, and use the compiler flags defined in this `makefile`, without changing the `makefile` of this project at all.

I don't understand the following design decisions of this library:
1. Instead of just having one `makefile`, we have `makefile` and `compile.py`, and `compile.py` actually calls `make`
2. Why do we need to have a centralized repository for plotting of all analyses? Isn't it better if we organize plots by analysis? For example, we can have layout like this:

```
lhcb-ntuples-gen
   |----ntuples
   |----lib/cpp/plot_scripts
   |----plots
       |----some_random_plot.cxx
       |----some_other_plot.cxx

rdx-run2-analysis
   |----ntuples
   |----lib/cpp/plot_scripts
   |----plots
       |----some_plot_for_ana_note.cxx
```

**Comment by yipengsun** [comment]
Of course, making this library integrable to other C++ project doesn't mean that this can't be used independently.

**Comment by yipengsun** [comment]
From the README of the mother project: https://github.com/richstu/ra4_draw, I see that you have analysis-specific instructions.

The `plot_scripts` serves as the central plot generation for all analysis. I don't like the idea (I prefer self-contained analysis), but it's consistent.

However, if we have all analysis in a single repo (hypothetically, J/Psi and lepton universality in a single repo), then we don't satisfy LHCb's analysis preservation requirement, so I think we should make analysis self-contained.

Well, not really self-contained, because the DaVinci scripts are not part of the analysis, but if we commit the step-2 ntuples with the analysis, then we'll at least satisfy LHCb's preservation requirement, though the ntuples will not be reproducible just with the preserved code.
